src/effects/deeplearning/EffectDeepLearning.cpp

```diff
@@ -0,0 +1,380 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+   License: GPL v2.  See License.txt.
+
+   EffectDeepLearning.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "EffectDeepLearning.h"
+#include "DeepModelManager.h"
+#include "ModelManagerPanel.h"
+
+#include "FileNames.h"
+#include "Shuttle.h"
+#include "ShuttleGui.h"
+#include <wx/range.h>
+
+#include <torch/script.h>
+
+#include <wx/log.h>
+#include <wx/stattext.h>
+
+#include <WaveClip.h>
+
+// ModelCardPanel
+
+EffectDeepLearning::EffectDeepLearning()
+{
+   mManagerPanel = nullptr;
+   mCard = nullptr;
+
+   EnablePreview(false);
+}
+
+bool EffectDeepLearning::Init()
+{
+   DeepModelManager &manager = DeepModelManager::Get();
+
+   // try loading the model (if available)
+   mModel = std::make_shared<DeepModel>();
+   if (mCard)
+   {
+      if (manager.IsInstalled(mCard))
+         mModel = manager.GetModel(mCard);
+   }
+
+   return true;
+}
+
+void EffectDeepLearning::End()
+{
+   DeepModelManager &manager = DeepModelManager::Get();
+
+   // release model (may still be active in thread)
+   mModel.reset();
+
+   // clean up in-progress installs
+   for (auto card : manager.GetCards())
+   {
+      if (manager.IsInstalling(card))
+         manager.CancelInstall(card);
+   }
+}
+
+bool EffectDeepLearning::Process()
+{
+   // throw an error if there isn't a model loaded
+   if (!mModel->IsLoaded())
+   {
+      Effect::MessageBox(
+          XO("Please install the selected model before applying the effect."),
+          wxICON_ERROR);
+      return false;
+   }
+
+   // Iterate over each track.
+   // All needed because this effect needs to introduce
+   // silence in the sync-lock group tracks to keep sync
+   CopyInputTracks(true); // Set up mOutputTracks.
+   bool bGoodResult = true;
+
+   mCurrentTrackNum = 0;
+
+   // NOTE: because we will append the separated tracks to mOutputTracks in ProcessOne(),
+   // we need to collect the track pointers before calling ProcessOne()
+   std::vector<WaveTrack *> pOutLeaders;
```

Comment:
yes! The reason I do this is because `mOutputTracks->SelectedLeaders<WaveTrack>()` is an iterator, and source separation modifies `mOutputTracks` while it's iterating, meaning that we will then feed the output tracks back into the loop (in a possibly infinite loop)